scvi/model/base/_pyromixin.py

```diff
@@ -0,0 +1,415 @@
+    """
+
+    def train(
+        self,
+        max_epochs: Optional[int] = None,
+        use_gpu: Optional[Union[str, int, bool]] = None,
+        train_size: float = 0.9,
+        validation_size: Optional[float] = None,
+        batch_size: int = 128,
+        early_stopping: bool = False,
+        plan_kwargs: Optional[dict] = None,
+        **trainer_kwargs,
+    ):
+        """
+        Train the model.
+
+        Parameters
+        ----------
+        max_epochs
+            Number of passes through the dataset. If `None`, defaults to
+            `np.min([round((20000 / n_cells) * 400), 400])`
+        use_gpu
+            Use default GPU if available (if None or True), or index of GPU to use (if int),
+            or name of GPU (if str, e.g., `'cuda:0'`), or use CPU (if False).
+        train_size
+            Size of training set in the range [0.0, 1.0].
+        validation_size
+            Size of the test set. If `None`, defaults to 1 - `train_size`. If
+            `train_size + validation_size < 1`, the remaining cells belong to a test set.
+        batch_size
+            Minibatch size to use during training.
+        early_stopping
+            Perform early stopping. Additional arguments can be passed in `**kwargs`.
+            See :class:`~scvi.train.Trainer` for further options.
+        plan_kwargs
+            Keyword args for :class:`~scvi.train.TrainingPlan`. Keyword arguments passed to
+            `train()` will overwrite values present in `plan_kwargs`, when appropriate.
+        **trainer_kwargs
+            Other keyword args for :class:`~scvi.train.Trainer`.
+        """
+        if max_epochs is None:
+            n_obs = self.adata.n_obs
+            max_epochs = np.min([round((20000 / n_obs) * 1000), 1000])
+
+        plan_kwargs = plan_kwargs if isinstance(plan_kwargs, dict) else dict()
+
+        if batch_size is None:
+            # use data splitter which moves data to GPU once
+            data_splitter = DeviceBackedDataSplitter(
+                self.adata,
+                train_size=train_size,
+                validation_size=validation_size,
+                batch_size=batch_size,
+                use_gpu=use_gpu,
+            )
+        else:
+            data_splitter = DataSplitter(
+                self.adata,
+                train_size=train_size,
+                validation_size=validation_size,
+                batch_size=batch_size,
+                use_gpu=use_gpu,
+            )
+        training_plan = PyroTrainingPlan(pyro_module=self.module, **plan_kwargs)
+
+        es = "early_stopping"
+        trainer_kwargs[es] = (
+            early_stopping if es not in trainer_kwargs.keys() else trainer_kwargs[es]
+        )
+        runner = TrainRunner(
+            self,
+            training_plan=training_plan,
+            data_splitter=data_splitter,
+            max_epochs=max_epochs,
+            use_gpu=use_gpu,
+            **trainer_kwargs,
```

Comment:
we should be able to detect if JIT trace elbo is being used here through the plan kwargs and add the callback accordingly.

I think if you look at the peakvi train method some callbacks are added as an example

https://github.com/YosefLab/scvi-tools/blob/a9da914764d311ff18b4b557b78d9550a3b940af/scvi/model/_peakvi.py#L216-L221

Reply:
@adamgayoso Would it hurt to always to this callback? (see what I added in https://github.com/YosefLab/scvi-tools/pull/1059/commits/cb3cae91fc6bab18b67b193520decbc8f526aeb6)

Reply:
Actually the callback does not work. I probably do not understand how to get dataloader from DataSplitter.

```python
========================================================================= short test summary info ==========================================================================
FAILED tests/models/test_pyro.py::test_pyro_bayesian_train_sample_mixin - AttributeError: 'DataSplitter' object has no attribute 'train_idx'
FAILED tests/models/test_pyro.py::test_pyro_bayesian_train_sample_mixin_full_data - AttributeError: 'DeviceBackedDataSplitter' object has no attribute 'train_tensor_dict'
FAILED tests/models/test_pyro.py::test_pyro_bayesian_train_sample_mixin_with_local - AttributeError: 'DataSplitter' object has no attribute 'train_idx'
FAILED tests/models/test_pyro.py::test_pyro_bayesian_train_sample_mixin_with_local_full_data - AttributeError: 'DeviceBackedDataSplitter' object has no attribute 'train_...
```

Reply:
yeah this is fine, just need to add `datasplitter.setup()` before trying to get the train data loadeer. Subsequent calls to setup are a no-op so this is fine.